### PR TITLE
Proper error reporting for connect failures

### PR DESCRIPTION
**Pull request by mnunberg**

### CODE REVIEW

**Comment by mnunberg** [comment]
Note that another PR (https://github.com/redis/hiredis/pull/461) also exists. Unlike the linked PR, this PR doesn't add a new API, and it doesn't break the abstraction layer between socket and context.

**Comment by mnunberg** [comment]
I'd also like to abstract getaddrinfo to be non-blocking; right now it's blocking and in async mode we never really rotate. I'd also like to abstract the sockaddr member as well.

**Comment by mnunberg** [comment]
@dvirsky

**Comment by dvirsky** [comment]
Cc: @yossigo can you review this?

File: net.c [hunk 1]
```diff
@@ -375,6 +399,15 @@ static int _redisContextConnectTcp(redisContext *c, const char *addr, int port,
                 goto error;
             }
         }
+
+        /* For repeat connection */
+        if (c->saddr) {
+            free(c->saddr);
+        }
+        c->saddr = malloc(sizeof(*p->ai_addr));
```
Comment (yossigo) [inline net.c]: The allocation should be done based on `ai_addrlen`, as `ai_addr` is a vanilla sockaddr which may be smaller than what you really need (e.g. for AF_UNIX or AF_INET6).

Comment (mnunberg) [reply net.c]: oops - done. There's also `struct sockaddr_storage`, but ideally I'd like to preserve the original `addrinfo` structure anyway.

**Comment by mnunberg** [comment]
Still working on getting the tests to pass..

**Pull request merged**